ui_framework/src/components/code_editor/code_editor.js

```diff
@@ -0,0 +1,91 @@
+import React, { Component } from 'react';
+import classNames from 'classnames';
+import AceEditor from 'react-ace';
+
+import { htmlIdGenerator, keyCodes } from '../../../services';
+
+export class KuiCodeEditor extends Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      hintInactive: false
+    };
+    this.idGenerator = htmlIdGenerator();
+  }
+
+  enableOverlay() {
+    this.setState({ hintInactive: false });
+  }
+
+  onKeydownAce = (ev) => {
+    if (ev.keyCode === keyCodes.ESCAPE) {
+      ev.preventDefault();
+      ev.stopPropagation();
+      this.enableOverlay();
+      this.refs.kuiCodeEditorHint.focus();
+    }
+  }
+
+  configureAce = (ace) => {
+    if (ace) {
+      this.aceEditor = ace;
+      ace.editor.textInput.getElement().tabIndex = -1;
+      ace.editor.textInput.getElement().addEventListener('keydown', this.onKeydownAce);
+    }
+  };
+
+  onBlurAce = (...args) => {
+    this.enableOverlay();
+    if (this.props.onBlur) {
+      this.props.onBlur(...args);
+    }
+  };
+
+  onHintKeyDown = (ev) => {
+    if (ev.keyCode === keyCodes.ENTER) {
+      ev.preventDefault();
+      this.startEditing();
+    }
+  };
+
+  startEditing = () => {
+    this.setState({ hintInactive: true });
+    this.aceEditor.editor.textInput.focus();
+  }
+
+  render() {
+    const { width, height } = this.props;
+    const classes = classNames('kuiCodeEditorKeyboardHint', {
+      'kuiCodeEditorKeyboardHint-isInactive': this.state.hintInactive
+    });
+    return (
+      <div
+        className="kuiCodeEditorWrapper"
+        style={{ width, height }}
+      >
+        <div
+          className={classes}
+          id={this.idGenerator('codeEditor')}
+          ref="kuiCodeEditorHint"
+          tabIndex="0"
+          role="button"
+          onClick={this.startEditing}
+          onKeyDown={this.onHintKeyDown}
+        >
+          <p className="kuiText kuiVerticalRhythmSmall">
+            Press Enter to start editing.
+          </p>
+          <p className="kuiText kuiVerticalRhythmSmall">
+            When you&rsquo;re done, press Escape to stop editing.
+          </p>
+        </div>
+        <AceEditor
+          {...this.props}
```

Comment:
I also thought about this, but decided against it for the following reasons (that are of course up for discussion):

* `AceEditor` has [quite some properties](https://github.com/securingsincity/react-ace/blob/master/docs/Ace.md#available-props), that we would need to extract and pass down.
* It also passes down all additional props to the ace editor itself. Especially for the mentioned `aria` attributes it would imho make more sense to apply them to the editor, than the wrapper element (that is never focused).

I totally see the point with the testing `data-test-subj` that you want most likely to be applied to the wrapper element. We have now several options:

1. We leave it as it is, and pass all properties to the editor, and just extract the ones we need for the wrapper element (which could also include the `data-test-subj`).
2. We pass by default all properties down to the editor, and just extract the ones we need (like `width` and `height`) and offer a specific property (e.g. `wrapperProps`), that you can pass arbitrary properties to the wrapper element.
3. We explicitly extract and pass all supported properties to `AceEditor` and all others to the wrapper (thus taking us the possibility to add additional (e.g. ARIA) attributes to the editor).
4. We explicitly extract properties for the editor, but offer an additional `editorProps` property, which will be assigned to the editor instead of the wrapper element.

Which way would you prefer?

Reply:
Reading it again after a day, I guess I would prefer 2. at the moment. That way we just pull out what we need, but still give the user a general way of attaching any property wherever she wants (using `wrapperProps` for the wrapper, or just applying them directly for the editor).

Reply:
Would 2 be also okay with you?

Reply:
After discussion on Zoom, we've decided to go with #1.